agent/auto-config/auto_config_test.go

```diff
@@ -116,99 +128,64 @@ func TestNew(t *testing.T) {
-	},
-		config.FileSource{
-			Name:   "overrides",
-			Format: "json",
-			Data:   `{"check_reap_interval": "1ms"}`,
-		})
-
-	require.NoError(t, err)
-	require.Empty(t, warnings)
-	require.NotNil(t, cfg)
-	require.Equal(t, "hobbiton", cfg.NodeName)
-	require.Equal(t, 1*time.Millisecond, cfg.CheckReapInterval)
-}
-
 func TestReadConfig(t *testing.T) {
 	// just testing that some auto config source gets injected
-	devMode := true
 	ac := AutoConfig{
 		autoConfigSource: config.LiteralSource{
 			Name:   autoConfigFileName,
 			Config: config.Config{NodeName: stringPointer("hobbiton")},
 		},
-		builderOpts: config.BuilderOpts{
-			// putting this in dev mode so that the config validates
-			// without having to specify a data directory
-			DevMode: &devMode,
-		},
 		logger: testutil.Logger(t),
+		acConfig: Config{
+			Loader: func(source config.Source) (*config.RuntimeConfig, []string, error) {
+				cfg, _, err := source.Parse()
+				if err != nil {
+					return nil, nil, err
+				}
+				return &config.RuntimeConfig{
+					DevMode:  true,
+					NodeName: *cfg.NodeName,
+				}, nil, nil
+			},
+		},
 	}
 
 	cfg, err := ac.ReadConfig()
 	require.NoError(t, err)
 	require.NotNil(t, cfg)
 	require.Equal(t, "hobbiton", cfg.NodeName)
+	require.True(t, cfg.DevMode)
 	require.Same(t, ac.config, cfg)
 }
 
-func testSetupAutoConf(t *testing.T) (string, string, config.BuilderOpts) {
+func setupRuntimeConfig(t *testing.T) *config.RuntimeConfig {
 	t.Helper()
 
-	// create top level directory to hold both config and data
-	tld := testutil.TempDir(t, "auto-config")
-	t.Cleanup(func() { os.RemoveAll(tld) })
-
-	// create the data directory
-	dataDir := filepath.Join(tld, "data")
-	require.NoError(t, os.Mkdir(dataDir, 0700))
+	dataDir := testutil.TempDir(t, "auto-config")
+	t.Cleanup(func() { os.RemoveAll(dataDir) })
 
-	// create the config directory
-	configDir := filepath.Join(tld, "config")
-	require.NoError(t, os.Mkdir(configDir, 0700))
-
-	builderOpts := config.BuilderOpts{
-		HCL: []string{
-			`data_dir = "` + dataDir + `"`,
-			`datacenter = "dc1"`,
-			`node_name = "autoconf"`,
-			`bind_addr = "127.0.0.1"`,
-		},
+	rtConfig := &config.RuntimeConfig{
+		DataDir:    dataDir,
+		Datacenter: "dc1",
+		NodeName:   "autoconf",
+		BindAddr:   &net.IPAddr{IP: net.ParseIP("127.0.0.1")},
 	}
-
-	return dataDir, configDir, builderOpts
+	return rtConfig
 }
 
 func TestInitialConfiguration_disabled(t *testing.T) {
-	dataDir, configDir, builderOpts := testSetupAutoConf(t)
-
-	cfgFile := filepath.Join(configDir, "test.json")
-	require.NoError(t, ioutil.WriteFile(cfgFile, []byte(`{
-		"primary_datacenter": "primary", 
-		"auto_config": {"enabled": false}
-	}`), 0600))
-
-	builderOpts.ConfigFiles = append(builderOpts.ConfigFiles, cfgFile)
-
-	directRPC := mockDirectRPC{}
-	conf := new(Config).
-		WithBuilderOpts(builderOpts).
-		WithDirectRPC(&directRPC)
+	rtConfig := setupRuntimeConfig(t)
+
+	directRPC := new(mockDirectRPC)
+	directRPC.Test(t)
```

Comment:
I don't know if there is an easier way to do this. When I caused some of the tests to fail they were panic-ing instead of failing. Setting `Mock.Test(t)`  fixed that problem.